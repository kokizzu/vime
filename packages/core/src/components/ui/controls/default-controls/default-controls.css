vime-default-controls {
  width: 100%;
}