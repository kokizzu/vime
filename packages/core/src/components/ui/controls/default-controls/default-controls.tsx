import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../../core/player/PlayerProp';

@Component({
  tag: 'vime-default-controls',
  styleUrl: 'default-controls.css',
})
export class DefaultControls {
  /**
   * The length in milliseconds that the controls are active for before fading out. Audio players
   * are not effected by this prop.
   */
  @Prop() activeDuration = 2750;

  /**
   * Whether the controls should wait for playback to start before being shown. Audio players
   * are not effected by this prop.
   */
  @Prop() waitForPlaybackStart = false;

  /**
   * Whether the controls should show/hide when paused. Audio players are not effected by this prop.
   */
  @Prop() hideWhenPaused = false;

  /**
   * Whether the controls should hide when the mouse leaves the player. Audio players are not
   * effected by this prop.
   */
  @Prop() hideOnMouseLeave = false;

  /**
   * @internal
   */
  @Prop() isMobile: PlayerProps[PlayerProp.IsMobile] = false;

  /**
   * @internal
   */
  @Prop() isLive: PlayerProps[PlayerProp.IsLive] = false;

  /**
   * @internal
   */
  @Prop() isAudioView: PlayerProps[PlayerProp.IsAudioView] = false;

  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.IsVideoView] = false;

  private buildAudioControls() {
    return (
      <vime-controls full-width>
        <vime-playback-control tooltip-direction="right" />
        <vime-volume-control />
        {!this.isLive && <vime-current-time />}
        {this.isLive && <vime-control-spacer />}
        {!this.isLive && <vime-scrubber-control /> }
        {this.isLive && <vime-live-indicator />}
        {!this.isLive && <vime-end-time />}
        {!this.isLive && <vime-settings-control tooltip-direction="left" />}
        <div style={{ paddingRight: '2px' }} />
      </vime-controls>
    );
  }

  private buildMobileVideoControls() {
    const lowerControls = (
      <vime-controls
        pin="bottomLeft"
        full-width
        activeDuration={this.activeDuration}
        waitForPlaybackStart={this.waitForPlaybackStart}
        hideWhenPaused={this.hideWhenPaused}
      >
        <vime-control-group>
          <vime-current-time />
          <vime-control-spacer />
          <vime-end-time />
          <vime-fullscreen-control />
        </vime-control-group>

        <vime-control-group space="top">
          <vime-scrubber-control />
        </vime-control-group>
      </vime-controls>
    );

    return (
      <Host>
        <vime-scrim />

        <vime-controls
          pin="topLeft"
          full-width
          activeDuration={this.activeDuration}
          waitForPlaybackStart={this.waitForPlaybackStart}
          hideWhenPaused={this.hideWhenPaused}
        >
          <vime-control-spacer />
          <vime-volume-control />
          {!this.isLive && <vime-caption-control />}
          {!this.isLive && <vime-settings-control />}
          {this.isLive && <vime-fullscreen-control />}
        </vime-controls>

        <vime-controls
          pin="center"
          activeDuration={this.activeDuration}
          waitForPlaybackStart={this.waitForPlaybackStart}
          hideWhenPaused={this.hideWhenPaused}
        >
          <vime-playback-control scale={1.5} />
        </vime-controls>

        {!this.isLive && lowerControls}
      </Host>
    );
  }

  private buildDesktopVideoControls() {
    const scrubberControlGroup = (
      <vime-control-group>
        <vime-scrubber-control />
      </vime-control-group>
    );

    return (
      <Host>
        <vime-scrim gradient="up" />

        <vime-controls
          activeDuration={this.activeDuration}
          waitForPlaybackStart={this.waitForPlaybackStart}
          hideWhenPaused={this.hideWhenPaused}
          hideOnMouseLeave={this.hideOnMouseLeave}
          full-width
        >
          {!this.isLive && scrubberControlGroup}

          <vime-control-group space={this.isLive ? 'none' : 'top'}>
            <vime-playback-control tooltip-direction="right" />
            <vime-volume-control />
            {!this.isLive && <vime-time-progress />}
            <vime-control-spacer />
            {!this.isLive && <vime-caption-control />}
            {this.isLive && <vime-live-indicator />}
            <vime-pip-control />
            {!this.isLive && <vime-settings-control />}
            <vime-fullscreen-control tooltip-direction="left" />
          </vime-control-group>
        </vime-controls>
      </Host>
    );
  }

  render() {
    if (this.isAudioView) return this.buildAudioControls();
    if (this.isVideoView && this.isMobile) return this.buildMobileVideoControls();
    if (this.isVideoView) return this.buildDesktopVideoControls();
    return undefined;
  }
}

openPlayerWormhole(DefaultControls, [
  PlayerProp.IsMobile,
  PlayerProp.IsAudioView,
  PlayerProp.IsVideoView,
  PlayerProp.IsLive,
]);
