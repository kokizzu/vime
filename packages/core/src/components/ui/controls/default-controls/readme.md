# vime-default-controls

Default set of controls for when you're in a hurry. The controls displayed depend on whether
the media is audio/video/live, and whether the device is mobile/desktop. See
[`vime-default-ui`](../../default-ui/readme.md) for visuals.

<!-- Auto Generated Below -->

## Usage

### Angular

```html {5-8} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-default-controls
      full-width
      active-duration="2750"
    ></vime-default-controls>
  </vime-ui>
</vime-player>
```

### Html

```html {5-8}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-default-controls
      full-width
      active-duration="2750"
    ></vime-default-controls>
  </vime-ui>
</vime-player>
```

### React

```tsx {2,10}
import React from 'react';
import { VimePlayer, VimeUi, VimeDefaultControls } from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>
        {/* ... */}
        <VimeDefaultControls fullWidth activeDuration="3200" />
      </VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {6,12,18} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
      <VimeDefaultControls activeDuration="3200" />
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi, VimeDefaultControls } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
      VimeDefaultControls,
    },
  };
</script>
```

## Properties

| Property               | Attribute                 | Description                                                                                                                 | Type      | Default |
| ---------------------- | ------------------------- | --------------------------------------------------------------------------------------------------------------------------- | --------- | ------- |
| `activeDuration`       | `active-duration`         | The length in milliseconds that the controls are active for before fading out. Audio players are not effected by this prop. | `number`  | `2750`  |
| `hideOnMouseLeave`     | `hide-on-mouse-leave`     | Whether the controls should hide when the mouse leaves the player. Audio players are not effected by this prop.             | `boolean` | `false` |
| `hideWhenPaused`       | `hide-when-paused`        | Whether the controls should show/hide when paused. Audio players are not effected by this prop.                             | `boolean` | `false` |
| `waitForPlaybackStart` | `wait-for-playback-start` | Whether the controls should wait for playback to start before being shown. Audio players are not effected by this prop.     | `boolean` | `false` |

## Dependencies

### Used by

- [vime-default-ui](../../default-ui)

### Depends on

- [vime-controls](../controls)
- [vime-playback-control](../playback-control)
- [vime-volume-control](../volume-control)
- [vime-current-time](../../time/current-time)
- [vime-control-spacer](../control-spacer)
- [vime-scrubber-control](../scrubber-control)
- [vime-live-indicator](../../live-indicator)
- [vime-end-time](../../time/end-time)
- [vime-settings-control](../settings-control)
- [vime-control-group](../control-group)
- [vime-fullscreen-control](../fullscreen-control)
- [vime-scrim](../../scrim)
- [vime-caption-control](../caption-control)
- [vime-time-progress](../../time/time-progress)
- [vime-pip-control](../pip-control)

### Graph

```mermaid
graph TD;
  vime-default-controls --> vime-controls
  vime-default-controls --> vime-playback-control
  vime-default-controls --> vime-volume-control
  vime-default-controls --> vime-current-time
  vime-default-controls --> vime-control-spacer
  vime-default-controls --> vime-scrubber-control
  vime-default-controls --> vime-live-indicator
  vime-default-controls --> vime-end-time
  vime-default-controls --> vime-settings-control
  vime-default-controls --> vime-control-group
  vime-default-controls --> vime-fullscreen-control
  vime-default-controls --> vime-scrim
  vime-default-controls --> vime-caption-control
  vime-default-controls --> vime-time-progress
  vime-default-controls --> vime-pip-control
  vime-playback-control --> vime-control
  vime-playback-control --> vime-icon
  vime-playback-control --> vime-tooltip
  vime-volume-control --> vime-mute-control
  vime-volume-control --> vime-slider
  vime-mute-control --> vime-control
  vime-mute-control --> vime-icon
  vime-mute-control --> vime-tooltip
  vime-current-time --> vime-time
  vime-scrubber-control --> vime-slider
  vime-scrubber-control --> vime-tooltip
  vime-end-time --> vime-time
  vime-settings-control --> vime-control
  vime-settings-control --> vime-icon
  vime-settings-control --> vime-tooltip
  vime-fullscreen-control --> vime-control
  vime-fullscreen-control --> vime-icon
  vime-fullscreen-control --> vime-tooltip
  vime-caption-control --> vime-control
  vime-caption-control --> vime-icon
  vime-caption-control --> vime-tooltip
  vime-time-progress --> vime-current-time
  vime-time-progress --> vime-end-time
  vime-pip-control --> vime-control
  vime-pip-control --> vime-icon
  vime-pip-control --> vime-tooltip
  vime-default-ui --> vime-default-controls
  style vime-default-controls fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
