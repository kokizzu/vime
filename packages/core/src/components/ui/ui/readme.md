# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Dependencies

### Used by

 - [vime-default-ui](../default-ui)

### Graph
```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
