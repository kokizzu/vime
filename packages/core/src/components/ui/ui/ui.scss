vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}