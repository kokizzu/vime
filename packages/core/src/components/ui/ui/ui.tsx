import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.isVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.isFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  PlayerProp.isVideoView,
  PlayerProp.playsinline,
  PlayerProp.isFullscreenActive,
]);
