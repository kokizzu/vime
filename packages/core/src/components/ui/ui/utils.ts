import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
