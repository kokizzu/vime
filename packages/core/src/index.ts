import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
