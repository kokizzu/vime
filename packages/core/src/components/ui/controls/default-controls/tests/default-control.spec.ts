import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../../ui/tests';
import { DefaultControls } from '../default-controls';

let page: SpecPage;
let controls: HTMLVimeDefaultControlsElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [DefaultControls],
    '<vime-default-controls />',
  ));

  controls = page.root!.querySelector('vime-default-controls')!;
});

it('should be structurally sound when mobile video', async () => {
  controls.isMobile = true;
  controls.isVideoView = true;
  await page.waitForChanges();
  expect(controls).toMatchSnapshot();
});

it('should be structurally sound when mobile video + live stream', async () => {
  controls.isMobile = true;
  controls.isLive = true;
  controls.isVideoView = true;
  await page.waitForChanges();
  expect(controls).toMatchSnapshot();
});

it('should be structurally sound when desktop video', async () => {
  controls.isVideoView = true;
  await page.waitForChanges();
  expect(controls).toMatchSnapshot();
});

it('should be structurally sound when desktop video + live stream', async () => {
  controls.isVideoView = true;
  controls.isLive = true;
  await page.waitForChanges();
  expect(controls).toMatchSnapshot();
});

it('should be structurally sound when audio', async () => {
  controls.isAudioView = true;
  await page.waitForChanges();
  expect(controls).toMatchSnapshot();
});

it('should be structurally sound when audio + live stream', async () => {
  controls.isAudioView = true;
  controls.isLive = true;
  await page.waitForChanges();
  expect(controls).toMatchSnapshot();
});
