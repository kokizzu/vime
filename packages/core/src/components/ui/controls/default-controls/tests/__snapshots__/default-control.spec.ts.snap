// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound when audio + live stream 1`] = `
<vime-default-controls>
  <vime-controls full-width="">
    <vime-playback-control tooltip-direction="right"></vime-playback-control>
    <vime-volume-control></vime-volume-control>
    <vime-current-time></vime-current-time>
    <vime-control-spacer></vime-control-spacer>
    <vime-live-indicator></vime-live-indicator>
    <div style="padding-right: 2px;"></div>
  </vime-controls>
</vime-default-controls>
`;

exports[`should be structurally sound when audio 1`] = `
<vime-default-controls>
  <vime-controls full-width="">
    <vime-playback-control tooltip-direction="right"></vime-playback-control>
    <vime-volume-control></vime-volume-control>
    <vime-current-time></vime-current-time>
    <vime-scrubber-control></vime-scrubber-control>
    <vime-end-time></vime-end-time>
    <vime-settings-control tooltip-direction="left"></vime-settings-control>
    <div style="padding-right: 2px;"></div>
  </vime-controls>
</vime-default-controls>
`;

exports[`should be structurally sound when desktop video + live stream 1`] = `
<vime-default-controls>
  <vime-scrim gradient="up"></vime-scrim>
  <vime-controls activeduration="2750" full-width="">
    <vime-control-group space="none">
      <vime-playback-control tooltip-direction="right"></vime-playback-control>
      <vime-volume-control></vime-volume-control>
      <vime-control-spacer></vime-control-spacer>
      <vime-live-indicator></vime-live-indicator>
      <vime-pip-control></vime-pip-control>
      <vime-fullscreen-control tooltip-direction="left"></vime-fullscreen-control>
    </vime-control-group>
  </vime-controls>
</vime-default-controls>
`;

exports[`should be structurally sound when desktop video 1`] = `
<vime-default-controls>
  <vime-scrim gradient="up"></vime-scrim>
  <vime-controls activeduration="2750" full-width="">
    <vime-control-group>
      <vime-scrubber-control></vime-scrubber-control>
    </vime-control-group>
    <vime-control-group space="top">
      <vime-playback-control tooltip-direction="right"></vime-playback-control>
      <vime-volume-control></vime-volume-control>
      <vime-time-progress></vime-time-progress>
      <vime-control-spacer></vime-control-spacer>
      <vime-caption-control></vime-caption-control>
      <vime-pip-control></vime-pip-control>
      <vime-settings-control></vime-settings-control>
      <vime-fullscreen-control tooltip-direction="left"></vime-fullscreen-control>
    </vime-control-group>
  </vime-controls>
</vime-default-controls>
`;

exports[`should be structurally sound when mobile video + live stream 1`] = `
<vime-default-controls>
  <vime-scrim></vime-scrim>
  <vime-controls activeduration="2750" full-width="" pin="topLeft">
    <vime-control-spacer></vime-control-spacer>
    <vime-volume-control></vime-volume-control>
    <vime-fullscreen-control></vime-fullscreen-control>
  </vime-controls>
  <vime-controls activeduration="2750" pin="center">
    <vime-playback-control scale="1.5"></vime-playback-control>
  </vime-controls>
</vime-default-controls>
`;

exports[`should be structurally sound when mobile video 1`] = `
<vime-default-controls>
  <vime-scrim></vime-scrim>
  <vime-controls activeduration="2750" full-width="" pin="topLeft">
    <vime-control-spacer></vime-control-spacer>
    <vime-volume-control></vime-volume-control>
    <vime-caption-control></vime-caption-control>
    <vime-settings-control></vime-settings-control>
  </vime-controls>
  <vime-controls activeduration="2750" pin="center">
    <vime-playback-control scale="1.5"></vime-playback-control>
  </vime-controls>
  <vime-controls activeduration="2750" full-width="" pin="bottomLeft">
    <vime-control-group>
      <vime-current-time></vime-current-time>
      <vime-control-spacer></vime-control-spacer>
      <vime-end-time></vime-end-time>
      <vime-fullscreen-control></vime-fullscreen-control>
    </vime-control-group>
    <vime-control-group space="top">
      <vime-scrubber-control></vime-scrubber-control>
    </vime-control-group>
  </vime-controls>
</vime-default-controls>
`;
